### PR TITLE
Implement type for IP address with stringer interface plus tests

### PR DESCRIPTION
**Pull request by runnerdave**
Fixes [#70](https://github.com/anz-bank/go-course/issues/70) 

Review of colleague's PR #366 

#### Changes proposed in this PR:

 - Write tests for a function int the spec in Readme.md
 - Implement type for IPAddr that represents an IP address
 - Implement stringer interface for this type
 - Pass tests

### CODE REVIEW

File: 05_stringer/runnerdave/main_test.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+func TestFibo(t *testing.T) {
+	tables := []struct {
+		x IPAddr
+		n string
+	}{
+		{IPAddr{127, 0, 0, 1}, "127.0.0.1"},
+		{IPAddr{10, 0, 0, 1}, "10.0.0.1"},
+		{IPAddr{191, 255, 255, 255}, "191.255.255.255"},
```
Comment (n0npax) [inline 05_stringer/runnerdave/main_test.go:32]: good job mate. Just one minor nit. Could you add test for empty and half initialized IPAddr
`IPAddr{10,10}, "10.10.0.0"`
`IPAddr{}, "0.0.0.0"`

**Review by n0npax — COMMENTED** [review]
Good job. 1 minor thing and I can approve. Cheers!

File: 05_stringer/runnerdave/main_test.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+func TestMainOutput(t *testing.T) {
+	var buf bytes.Buffer
+	out = &buf
+
+	main()
+
+	expected := strconv.Quote("127.0.0.1\n")
+	actual := strconv.Quote(buf.String())
+	t.Logf("expected:%s", expected)
```
Comment (brnbke) [inline 05_stringer/runnerdave/main_test.go:17]: More of a question/comment than a change, but in this test you log the expected and actual always, and maybe an error, on your other test you only log when there's an error. Expending this out if you have a lot of tests, would it get a bit noisy to log during success criteria?

Comment (runnerdave) [reply 05_stringer/runnerdave/main_test.go:17]: hi @brnbke this reflects my TDD approach, note that the logging is only in the single test not on the table one, because I write a failing test and then write the function to pass it, if I added logging in the other test I agree the noise would not be useful at all.

File: 05_stringer/runnerdave/main_test.go (hunk 1 above)
Comment (runnerdave) [reply 05_stringer/runnerdave/main_test.go:32]: implemented, thanks for the input

**Review by n0npax — DISMISSED** [review]
lgtm

File: 05_stringer/runnerdave/main_test.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+	t.Logf("actual:%s", actual)
+
+	if expected != actual {
+		t.Errorf("Unexpected output in main()")
+	}
+}
+
+func TestFibo(t *testing.T) {
```
Comment (pentaphobe) [inline 05_stringer/runnerdave/main_test.go]: Is it really testing fibonacci? ;)

Totally a nitpick, but good to watch out for such things when copy-pasting especially if they're publicly exported.

(Or totally avoid copy-pasting when first acclimatising to a new language)

**Review by pentaphobe — DISMISSED** [review]
LGTM also

keep an eye out for incorrect copypasta though!

File: 05_stringer/runnerdave/main_test.go (hunk 3 above)
Comment (anzboi) [reply 05_stringer/runnerdave/main_test.go]: Please change this.

**Review by anzboi — DISMISSED** [review]

**Comment by anzboi** [comment]
Change the test function name then all good.

File: 05_stringer/runnerdave/main_test.go (hunk 3 above)
Comment (runnerdave) [reply 05_stringer/runnerdave/main_test.go]: hehe good catch

File: 05_stringer/runnerdave/main_test.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+		{IPAddr{10, 10}, "10.10.0.0"},
+		{IPAddr{}, "0.0.0.0"},
+	}
+
+	for _, table := range tables {
+		var buf bytes.Buffer
+		out = &buf
+		fmt.Fprintln(out, table.x)
+		expected := strconv.Quote(table.n)
+		actual := strconv.Quote(strings.Replace(buf.String(), "\n", "", -1))
```
Comment (camscale) [inline 05_stringer/runnerdave/main_test.go]: Using `Fprintln` to call the `String` method and then stripping off the `\n` added by `Fprintln` seems unnecessary. You could just call `Fprint` instead (https://golang.org/pkg/fmt/#Fprint) and not have a newline to strip off.

I'd also consider removing `Fprint(ln)` altogether and just call `table.x.String()`. Using a buffer to capture the output is required for `main`, but when testing a function with inputs and outputs, it makes more sense to deal directly with those inputs and outputs with nothing in between.

Finally, since there are no special characters in the string literals now, there is no need to use `strconv.Quote`.

With all this in place, your test should just be:
```
expected := table.n
actual := table.x.String()
```

That is much simpler to read an understand.

On the subject of reading and understanding, using `n` and `x` for the fields in your struct are rather meaningless and even misleading. `n` is usually some sort of count (int). I think naming them something like `input` (or `in`) and `expected` (or `out`) would make this more readable. Go encourages short variable names, but they need to be meaningful. In this context, `n` and `x` are not meaningful. They just seem to be random characters.

If you are unsure of naming, it is often useful to look at the Go library source code. E.g. https://golang.org/src/strconv/atob_test.go (the strconv package) uses `in` and `out`, so following that means you likely wont go wrong.

**Review by camscale — COMMENTED** [review]
Just a general comment on making tests simple/minimal and readable. Let me know if you disagree - I'll approve given a reasonable counter-argument.

File: 05_stringer/runnerdave/main_test.go (hunk 4 above)
Comment (runnerdave) [reply 05_stringer/runnerdave/main_test.go]: @camh-anz  this is really valuable feedback thanks a lot, will implement later

File: 05_stringer/runnerdave/main_test.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+		{IPAddr{191, 255, 255, 255}, "191.255.255.255"},
+		{IPAddr{10, 10}, "10.10.0.0"},
+		{IPAddr{}, "0.0.0.0"},
+	}
+
+	for _, table := range tables {
+		expected := table.output
+		actual := table.input.String()
+		if expected != actual {
+			t.Errorf("Stringer of (%d) was incorrect, got: %s, want: %s.", table.input, actual, expected)
```
Comment (camscale) [inline 05_stringer/runnerdave/main_test.go]: Sorry I didn't notice this before, but the `%d` format specifier here is incorrect. I would suggest `%#v` which will give a Go-syntax representation of the value (https://golang.org/pkg/fmt/) as shown here: https://goplay.space/#BiRitf92ruM (`%v` would call the `String()` method which is counter-productive here)

**Review by camscale — CHANGES_REQUESTED** [review]
Sorry, one thing I missed last time...

File: 05_stringer/runnerdave/main_test.go (hunk 5 above)
Comment (runnerdave) [reply 05_stringer/runnerdave/main_test.go]: implemented

**Review by camscale — APPROVED** [review]
LGM. Thanks for your patience working through this. Looks great now.
